client/reducers/runs.js

```diff
@@ -0,0 +1,117 @@
+            return {
+                ...state,
+                activeRunsById: runs,
+                activeRunConfiguration: config
+            };
+        }
+        case DELETE_USERS_FROM_RUN: {
+            const { usersForRun, runId } = action.payload;
+            const runToUpdate = state.activeRunsById[runId];
+            return {
+                ...state,
+                activeRunsById: {
+                    ...state.activeRunsById,
+                    [runId]: {
+                        ...runToUpdate,
+                        testers: usersForRun
+                    }
+                }
+            };
+        }
+        case SAVE_RUN_STATUS: {
+            const { run_status, run_status_id, runId } = action.payload;
+
+            const oldStatus = state.activeRunsById[runId].run_status;
+            const updatedRun = {
+                ...state.activeRunsById[runId],
+                run_status,
+                run_status_id
+            };
+
+            if (oldStatus !== 'final' && run_status !== 'final') {
+                return {
+                    ...state,
+                    activeRunsById: {
+                        ...state.activeRunsById,
+                        [runId]: updatedRun
+                    }
+                };
+            } else if (oldStatus !== 'final' && run_status === 'final') {
+                const updatedActiveRunsById = { ...state.activeRunsById };
+                delete updatedActiveRunsById[runId];
+                let updatedPublishedRunsById = undefined;
+
+                // If we switched the run from active to published,
+                // Then we might not have fetched the published runs yet
+                if (state.publishedRuns) {
+                    updatedPublishedRunsById = {
+                        ...this.state.publishedRunsById,
+                        [runId]: updatedRun
+                    };
+                }
+                return {
+                    ...state,
+                    activeRunsById: updatedActiveRunsById,
+                    publishedRunsById: updatedPublishedRunsById
+                };
+            } else if (oldStatus === 'final' && run_status !== 'final') {
+                const updatedPublishedRunsById = { ...state.publishedRunsById };
+                delete updatedPublishedRunsById[runId];
+                let updatedActiveRunsById = undefined;
+
+                // If we switched the run from published to active,
+                // Then we might not have fetched the active runs yet
+                if (state.activeRunsById) {
+                    updatedActiveRunsById = {
+                        ...this.state.activeRunsById,
+                        [runId]: updatedRun
+                    };
+                }
+                return {
+                    ...state,
+                    activeRunsById: updatedActiveRunsById,
+                    publishedRunsById: updatedPublishedRunsById
+                };
```

Comment:
I'm thinking there should be a default case here for a clean return:

```
return { ...state }
```

Reply:
good idea